jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java

```diff
@@ -150,37 +151,51 @@ public Content.Chunk read()
         if (tunnelSupport != null)
             return null;
 
-        while (true)
+        // Check if there already is a chunk, e.g. EOF.
+        Content.Chunk chunk;
+        try (AutoLock ignored = lock.lock())
         {
-            Content.Chunk chunk;
-            try (AutoLock ignored = lock.lock())
-            {
-                chunk = _chunk;
-                _chunk = Content.Chunk.next(chunk);
-            }
-            if (chunk != null)
-                return chunk;
-
-            Stream.Data data = _stream.readData();
-            if (data == null)
-                return null;
-
-            // The data instance should be released after readData() above;
-            // the chunk is stored below for later use, so should be retained;
-            // the two actions cancel each other, no need to further retain or release.
-            chunk = createChunk(data);
-
-            // Some content is read, but the 100 Continue interim
-            // response has not been sent yet, then don't bother
-            // sending it later, as the client already sent the content.
-            if (_expects100Continue && chunk.hasRemaining())
-                _expects100Continue = false;
+            chunk = _chunk;
+            _chunk = Content.Chunk.next(chunk);
+        }
+        if (chunk != null)
+            return chunk;
+
+        Stream.Data data = _stream.readData();
+        if (data == null)
+            return null;
 
+        // Check if the trailers must be returned.
+        if (data.frame().isEndStream())
+        {
+            Content.Chunk trailer;
             try (AutoLock ignored = lock.lock())
             {
-                _chunk = chunk;
+                trailer = _trailer;
```

Comment:
You could also null out `_trailer` here.

Reply:
As `HttpStreamOverHTTP2` instances are per-request, I prefer to not null out the field, so that in case of dump/memory inspection we will know if a trailer was delivered.